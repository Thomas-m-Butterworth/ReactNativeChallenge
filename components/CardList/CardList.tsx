import React from "react";
import { FlatList, StyleSheet, View } from "react-native";
import { useDataStore } from "../../store/data/dataStore";
import { Card } from "../../components/Card/Card";
import { CardProps } from "../../store/data/types";

export const CardList = ({ data }: { data: CardProps[] }) => {
  return (
    <View style={styles.container}>
      <FlatList
        data={data}
        keyExtractor={(item, index) => `${item.from}-${index}`}
        renderItem={({ item }) => <Card card={item} />}
        contentContainerStyle={styles.listContent}
        showsVerticalScrollIndicator={false}
        testID="CardList"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  listContent: {
    paddingBottom: 16,
  },
});
