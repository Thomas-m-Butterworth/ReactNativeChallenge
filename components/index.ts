export * from "./Card";
export * from "./CardList";
export * from "./Metrics";
export * from "./Time";
