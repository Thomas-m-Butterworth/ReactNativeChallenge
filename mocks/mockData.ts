import { CardProps } from "../store/data/types";

export const mockData: CardProps[] = [
  {
    from: "2024-01-20T14:00Z",
    to: "2024-01-20T14:30Z",
    intensity: {
      forecast: 45,
      actual: 42,
      index: "very-low",
    },
  },
  {
    from: "2024-01-20T14:30Z",
    to: "2024-01-20T15:00Z",
    intensity: {
      forecast: 89,
      actual: 85,
      index: "low",
    },
  },
];
