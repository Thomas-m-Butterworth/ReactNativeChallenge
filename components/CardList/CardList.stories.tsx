import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import { View } from "react-native";
import { CardList } from "./CardList";
import { mockData } from "../../mocks/mockData";

const meta = {
  title: "CardList",
  component: CardList,
  decorators: [
    (Story) => (
      <View style={{ flex: 1, padding: 16 }}>
        <Story />
      </View>
    ),
  ],
} satisfies Meta<typeof CardList>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    data: mockData,
  },
};
