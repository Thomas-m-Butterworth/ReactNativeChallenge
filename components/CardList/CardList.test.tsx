import React from "react";
import { render } from "@testing-library/react-native";
import { CardList } from "../../components/CardList";
import { mockData } from "../../mocks/mockData";

describe("CardList", () => {
  it("should render a FlatList with the correct number of items", () => {
    const { getAllByTestId, getByTestId } = render(
      <CardList data={mockData} />
    );
    const cardList = getByTestId("CardList");
    expect(cardList).toBeTruthy();

    const cards = getAllByTestId("CardContainer");
    expect(cards.length).toBe(mockData.length);
  });

  it("should match the snapshot", () => {
    const { toJSON } = render(<CardList data={mockData} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
