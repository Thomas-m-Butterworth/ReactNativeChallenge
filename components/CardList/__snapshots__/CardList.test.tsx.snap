// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 16,
      }
    }
    data={
      [
        {
          "from": "2024-01-20T14:00Z",
          "intensity": {
            "actual": 42,
            "forecast": 45,
            "index": "very-low",
          },
          "to": "2024-01-20T14:30Z",
        },
        {
          "from": "2024-01-20T14:30Z",
          "intensity": {
            "actual": 85,
            "forecast": 89,
            "index": "low",
          },
          "to": "2024-01-20T15:00Z",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    testID="CardList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ffe4b5",
              "borderRadius": 16,
              "height": 150,
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 16,
            }
          }
          testID="CardContainer"
        >
          <View
            style={
              {
                "alignItems": "center",
                "marginBottom": 16,
              }
            }
            testID="TimeRangeContainer"
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "letterSpacing": -0.35,
                }
              }
            >
              14:00
               - 
              14:30
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginTop": 5,
              }
            }
            testID="MetricsContainer"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="ForecastContainer"
            >
              <FeatherMock
                color="#00bfff"
                name="trending-up"
                size={30}
                testID="ForecastIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="ForecastText"
              >
                45
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="ActualContainer"
            >
              <FeatherMock
                color="#3cb371"
                name="eye"
                size={30}
                testID="ActualIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="ActualText"
              >
                42
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="IndexContainer"
            >
              <FeatherMock
                color="#00bfff"
                name="bar-chart-2"
                size={30}
                testID="IndexIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="IndexText"
              >
                Very low
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ffe4b5",
              "borderRadius": 16,
              "height": 150,
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 16,
            }
          }
          testID="CardContainer"
        >
          <View
            style={
              {
                "alignItems": "center",
                "marginBottom": 16,
              }
            }
            testID="TimeRangeContainer"
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "letterSpacing": -0.35,
                }
              }
            >
              14:30
               - 
              15:00
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginTop": 5,
              }
            }
            testID="MetricsContainer"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="ForecastContainer"
            >
              <FeatherMock
                color="#00bfff"
                name="trending-up"
                size={30}
                testID="ForecastIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="ForecastText"
              >
                89
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="ActualContainer"
            >
              <FeatherMock
                color="#3cb371"
                name="eye"
                size={30}
                testID="ActualIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="ActualText"
              >
                85
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              testID="IndexContainer"
            >
              <FeatherMock
                color="#3cb371"
                name="bar-chart-2"
                size={30}
                testID="IndexIcon"
              />
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "marginTop": 8,
                  }
                }
                testID="IndexText"
              >
                Low
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
